File: Haskell/Menu/MenuLogado.hs
module Menu.MenuLogado where
import Controller.Usuario
import Controller.Livro
import Controller.Perfil
import Controller.Leitura
import Controller.Avaliacao
import Controller.Admin
import Controller.Estatisticas
import Menu.MenuEstatisticas
import Data.Maybe


exibeMenuLogado :: Usuario -> IO()
exibeMenuLogado usuario = do
    putStrLn $ "-------------------------------------" ++ "\n"
    putStrLn $ "|       Bem vindo ao The Readers     |" ++ "\n"
    putStrLn $ "-------------------------------------" ++ "\n"

    putStrLn "\n [P] Meu Perfil\n [F] Feed\n [U] Seguir usuário\n [B] Buscar usuário\n [L] Cadastrar Leitura\n [R] Criar Resenha\n [+] Cadastro de Livro\n [-] Excluir um livro\n [M] Minhas Estantes\n [E] Estatísticas\n [S] Sair"


    opcao <- getLine
    selecionaAcaoLogin usuario opcao

selecionaAcaoLogin :: Usuario -> String -> IO ()
selecionaAcaoLogin usuario "S" = do
    putStrLn "Obrigado"

selecionaAcaoLogin usuario "P" = do
     menuPerfil usuario

selecionaAcaoLogin usuario "+" = do
    putStrLn "Nome do livro: "
    nome <- getLine

    putStrLn "Autor: "
    autor <- getLine

    putStrLn "Número de páginas: "
    input <- getLine
    let n_paginas = read input :: Int

    putStrLn "Gênero: "
    genero <- getLine

    cadastraLivro nome autor n_paginas genero

selecionaAcaoLogin usuario "-" = do
    putStrLn "Nome do livro a ser excluido: "
    nomeLivro <- getLine
    deletaLivro nomeLivro

selecionaAcaoLogin usuario "B" = do
    menuPerfilStalker usuario

selecionaAcaoLogin usuario "A" = do
    let nomesUsuarios = recuperaNomeDeUsuarios recuperaUsuariosUnsafe []
    let nomesFiltrados = (removeElementos nomesUsuarios ([idUsuario usuario] ++ seguindo usuario))
    putStrLn "Usuários: "
    putStrLn ""
    imprimeLista nomesFiltrados
    putStrLn "Seguir usuário: "
    seguir <- getLine
    tentaSeguir usuario nomesFiltrados seguir

selecionaAcaoLogin usuario "F" = do
    maybeAvaliacoes <- recuperaAvaliacoesSafe
    let avaliacoes = fromJust maybeAvaliacoes
    seguindo <- recuperaSeguidores usuario
    let avaliacoesSeguindo = recuperaAvaliacoesPorNome (seguindo) avaliacoes
    simuladorFeed usuario avaliacoesSeguindo
    exibeMenuLogado usuario

selecionaAcaoLogin usuario "L" = do
    putStrLn "Nome do livro: "
    nomeLivro <- getLine
    putStrLn "Data da leitura: "
    dataLeitura <- getLine
    putStrLn "Nota da leitura (1 - 5)"
    nota <- readLn :: IO Int
    maybeListaLivros <- getListaLivros
    let listaLivros = fromMaybe [] maybeListaLivros
    let maybeLivro = getLivro nomeLivro listaLivros
    tentaCadastrar usuario maybeLivro dataLeitura nota

selecionaAcaoLogin usuario "E" = do
    putStrLn $ "-------------------------------------" ++ "\n"
    putStrLn $ "|            Estatísticas           |" ++ "\n"
    putStrLn $ "-------------------------------------" ++ "\n"
    menuEstatisticas usuario
    exibeMenuLogado usuario

selecionaAcaoLogin usuario "R" = do
    putStrLn "Qual leitura você deseja resenhar:"
    todasLeituras <- recuperaLeituraSafe
    case todasLeituras of
        Just leituras -> do 
            let lista_leituras = recuperaLeituraDoUsuarioSafe (idUsuario usuario) leituras 
            let nomes_das_leituras = concatenaLidos lista_leituras []

            imprimeLista nomes_das_leituras

            nomeLeitura <- getLine

            let maybeLeitura = procuraLeitura (idUsuario usuario) nomeLeitura (recuperaLeituraDoUsuarioSafe (idUsuario usuario) leituras)

            putStrLn "Escreva sua resenha: "
            resenha <- getLine
            
            tentaResenhar usuario maybeLeitura resenha
        Nothing -> putStrLn "Falha no JSON"
    


tentaResenhar :: Usuario -> Maybe Leitura -> String -> IO()
tentaResenhar usuario maybeLeitura resenha = 
    case maybeLeitura of
        Just leitura -> do
            fazAvaliacao (idUsuario usuario) leitura resenha
            putStrLn "Resenha cadastrada!"
            exibeMenuLogado usuario
        Nothing -> do
            putStrLn "Título não lido"
            exibeMenuLogado usuario



tentaCadastrar :: Usuario -> Maybe Livro -> String -> Int -> IO()
tentaCadastrar usuario maybeLivro dataLeitura nota = do
    case maybeLivro of 
        Just livro -> do
            if nota >= 1 && nota <= 5 then do
                cadastrarLeitura (idUsuario usuario) livro dataLeitura nota
                putStrLn "Livro cadastrado!"
                exibeMenuLogado usuario
            else do 
                putStrLn "Nota invalida"
                exibeMenuLogado usuario
        Nothing -> do
            putStrLn "Livro não encontrado no sistema"
            exibeMenuLogado usuario

tentaSeguir :: Usuario -> [[Char]] -> String -> IO()
tentaSeguir usuario validos seguir = do
    case (seguir `elem` validos) of
        True -> do
            atualizaSeguindo usuario seguir 
            putStrLn "Usuário seguido com sucesso!"
            exibeMenuLogado usuario
        False -> do 
            putStrLn "Usuário inválido"
            exibeMenuLogado usuario
    

imprimeLista :: [[Char]] -> IO()
imprimeLista [] = putStrLn ""
imprimeLista (x:xs) = do
     putStr "- " 
     putStrLn(x) 
     imprimeLista xs

removeElementos :: [String] -> [String] -> [String]
removeElementos listaTotal remover = [nome | nome <- listaTotal, not (nome `elem` remover)]

menuPerfil :: Usuario -> IO()
menuPerfil usuario = do
     putStrLn "\nEscolher opção: \n [V] Visão geral \n [E] Editar meu perfil\n [M] Minhas Resenhas\n [S] Voltar ao menu principal"
     opcao <- getLine
     selecionaOpcao usuario opcao


selecionaOpcao :: Usuario -> String -> IO ()

selecionaOpcao usuario "V" = do
    visaoGeral (idUsuario usuario) (seguidores usuario) (seguindo usuario)
    putStrLn $ "\n--------------------------------------" ++ "\n"
    menuPerfil usuario

selecionaOpcao usuario "E" = do
    putStrLn "Escolha seu nome: "
    nome <- getLine

    putStrLn "Escolha sua biografia: "
    biografia <- getLine

    criarPerfil nome biografia (idUsuario usuario)
    putStrLn "Perfil salvo com sucesso!"

    menuPerfil usuario

selecionaOpcao usuario "S" = do
    exibeMenuLogado usuario

selecionaOpcao usuario "M" = do
    carregaAvaliacoes usuario

selecionaOpcao usuario "" = do
     putStrLn "Opção Inválida"
     menuPerfil usuario



{-Carrega as Resenhas do Usuário Logado-}
carregaAvaliacoes :: Usuario -> IO()
carregaAvaliacoes usuario = do
    avaliacoes <- recuperaAvaliacoesSafe
    case avaliacoes of
        Just avaliacoes -> do
            let minhasAvaliacoes = recuperaAvaliacoesUsuario [usuario] avaliacoes
            putStrLn "----------- RESENHAS ----------"
            imprimeAvaliacoes minhasAvaliacoes
            menuPerfil usuario
        Nothing -> do
            putStrLn "Nenhuma avaliação"
            menuPerfil usuario

{-Itera imprimindo cada avaliação-}
imprimeAvaliacoes :: [Avaliacao] -> IO()
imprimeAvaliacoes [] = putStrLn "------------- FIM -------------"
imprimeAvaliacoes (x:xs) = do 
    let string = mostraAvaliacao x 
    putStrLn "-------------------------------"
    putStrLn string
    imprimeAvaliacoes xs

menuEstante :: Usuario -> IO()
menuEstante usuario = do
    putStrLn "\nEscolher opção: \n [+] Adicionar Livro \n [L] Lidos\n [E] Lendo\n [P] Pretendo Ler\n [A] Abandonados\n [S] Voltar para menu principal"
    opcao <- getLine
    selecionaOpcao usuario opcao    


menuPerfilStalker :: Usuario -> IO()
menuPerfilStalker usuario = do
     putStrLn "Digite o login do perfil que você deseja visitar:"
     userVisitado <- getLine
     visaoStalker userVisitado
     menuPerfil usuario

exibeDashAdm :: Admin -> IO()
exibeDashAdm adm = do
    putStrLn $ "-------------------------------------" ++ "\n"
    putStrLn $ "|     Dashboard de Administrador     |" ++ "\n"
    putStrLn $ "-------------------------------------" ++ "\n"

    putStrLn "\n [1] Estatísticas Gerais\n [2] Listar de usuários cadastrados\n [3] Listar livros cadastrados\n [+] Cadastrar novo adm\n [S] Sair"

    escolha <- getLine
    escolhaAdm adm escolha

escolhaAdm :: Admin -> String -> IO ()
escolhaAdm adm "1" = do
    exibeEstatisticasAdmin
    exibeDashAdm adm

escolhaAdm adm "2" = do
    listaUsers adm
    exibeDashAdm adm

escolhaAdm adm "3" = do
    listaLivros adm
    exibeDashAdm adm

escolhaAdm adm "+" = do
    putStrLn "Novo login administrador: "
    login <- getLine

    putStrLn "Escolha uma senha: "
    senha <- getLine
    cadastraAdm login senha
    putStrLn "Adm Cadastrado!"
    exibeDashAdm adm

escolhaAdm adm "S" = do
    putStrLn "Até a próxima!"

escolhaAdm adm "" = do
     putStrLn "Opção Inválida"
     exibeDashAdm adm

File: Prolog/Menu/MenuLogado.pl
:- module(menulogado, [menuLogado/1, menuPerfil/1, stalkear/1, editaPerfil/2 ]).
:- use_module("../Controller/Usuario.pl").
:- use_module("../Controller/Leitura.pl").
:- use_module("../Controller/Estante.pl").
:- use_module("../Controller/Perfil.pl").
:- use_module("../Controller/Estatisticas.pl").
:- use_module("../Controller/Resenha.pl").

menuLogado(Usuario):-
    nl,
    writeln("--------------------------------------"),
    writeln("|       Bem vindo ao The Readers     |"),
    writeln("--------------------------------------"),
    imprimeOpcoes(Opcao),
    selecionaAcao(Opcao, Usuario).

seguirUsuario(Usuario):- tty_clear, seguir(Usuario, UsuarioAtt), menuLogado(UsuarioAtt).

imprimeOpcoes(Opcao):-
    nl,
    nl, writeln("------------ M E N U : ------------"), nl,
    writeln("[F] Ir para o Feed"),
    writeln("[P] Meu Perfil"),
    writeln("[U] Seguir Usuário"),
    writeln("[M] Minhas Estantes"),
    writeln("[+] Cadastro de Livro"),
    writeln("[L] Cadastrar Leitura"),
    writeln("[R] Cadastrar Resenha"),
    writeln("[E] Estatisticas"),
    writeln("[S] Sair"),
    read_line_to_string(user_input, Opcao).

selecionaAcao(Opcao, Usuario):- (
                        Opcao == "S" -> nl, writeln("Obrigado por acessar o The Readers"), halt;
                        Opcao == "U" -> seguirUsuario(Usuario);
                        Opcao == "L" -> cadastrarLeitura(Usuario);
                        Opcao == "+" -> cadastraLivro, menuLogado(Usuario), !;
                        Opcao == "M" -> menuEstante(Usuario);
                        Opcao == "E" -> menuEstatisticas(Usuario);
                        Opcao == "R" -> cadastrarResenha(Usuario);
                        Opcao == "P" -> menuPerfil(Usuario);
                        Opcao == "F" -> feed(Usuario);
                        writeln("Ação inválida"), menu, !).


cadastraLivro:-
    writeln("Nome do livro: "),
    read_line_to_string(user_input, Nome),
    writeln("Autor: "),
    read_line_to_string(user_input, Autor),
    writeln("Número de páginas: "),
    read_line_to_string(user_input, Paginas_String),
    atom_number(Paginas_String, N_Paginas),
    writeln("Gênero: "),
    read_line_to_string(user_input, Genero),
    criaLivro(Nome, Autor, N_Paginas, Genero),
    writeln("").


menuPerfil(Usuario):-
    nl,
    writeln("Escolher opção:"),
    imprimeOpcoesPerfil(Opcao),
    selecionaAcaoPerfil(Opcao, Usuario).

imprimeOpcoesPerfil(Opcao):-
    nl,
    writeln("[V] Visão geral"),
    writeln("[E] Editar meu perfil"),
    writeln("[O] Visitar outro perfil"),
    writeln("[R] Minhas Resenhas"),
    writeln("[S] Voltar ao menu principal"),
    read_line_to_string(user_input, Opcao).

selecionaAcaoPerfil(Opcao, Usuario):- (
                        Opcao == "S" -> menuLogado(Usuario);
                        Opcao == "V" -> nl,
                                    writeln("--------------------------------------"),
                                    writeln("|        MEU PERFIL THE READER        |"),
                                    writeln("--------------------------------------"),
                                    visaoGeral(Usuario.nome), menuPerfil(Usuario), !;
                        Opcao == "E" -> editaPerfil(Usuario, Usuario.nome);
                        Opcao == "O" -> stalkear(Usuario);
                        Opcao == "R" -> minhasResenhas(Usuario), menuPerfil(Usuario), !;
                        writeln("Ação inválida"), menu, !).

editaPerfil(Usuario, NomeUsuario):- (
    nl,
    writeln("Escolha seu nome: "),
    read_line_to_string(user_input,NomePerfil),
    writeln("Escolha sua biografia: "),
    read_line_to_string(user_input,Biografia),
    criaPerfil(NomePerfil, Biografia, NomeUsuario),
    writeln("Perfil salvo com sucesso!"), menuPerfil(Usuario), !).


stalkear(Usuario):- (
    nl,
    writeln("Insira o ID do perfil à ser visitado: "),
    read_line_to_string(user_input, PerfilVisitado),
    writeln("--------------------------------------"),
    writeln("|        CONHEÇA ESSE READER :)      |"),
    writeln("--------------------------------------"),
    visaoStalker(PerfilVisitado), menuPerfil(Usuario), !).